Pass the current point to SVGPathSource segment parsing functions

smfr opened the pull request:
#### 75438506e981119be69203a2c709d2e02a8c5215
<pre>
Pass the current point to SVGPathSource segment parsing functions
<a href="https://bugs.webkit.org/show_bug.cgi?id=281645">https://bugs.webkit.org/show_bug.cgi?id=281645</a>
<a href="https://rdar.apple.com/138088365">rdar://138088365</a>

Reviewed by Alan Baradlay.

The CSS `shape()` function is getting start/end-relative control points[1] and so its SVGPathSource
implementation will need to know the current path point. So pass the current point to all SVGPathSource
functions that create non-trivial segments.

[1] <a href="https://github.com/w3c/csswg-drafts/issues/10649#issuecomment-2412816773">https://github.com/w3c/csswg-drafts/issues/10649#issuecomment-2412816773</a>

* Source/WebCore/rendering/style/BasicShapesShape.cpp:
* Source/WebCore/svg/SVGPathBlender.cpp:
(WebCore::pullFromSources):
(WebCore::SVGPathBlender::blendMoveToSegment):
(WebCore::SVGPathBlender::blendLineToSegment):
(WebCore::SVGPathBlender::blendLineToHorizontalSegment):
(WebCore::SVGPathBlender::blendLineToVerticalSegment):
(WebCore::SVGPathBlender::blendCurveToCubicSegment):
(WebCore::SVGPathBlender::blendCurveToCubicSmoothSegment):
(WebCore::SVGPathBlender::blendCurveToQuadraticSegment):
(WebCore::SVGPathBlender::blendCurveToQuadraticSmoothSegment):
(WebCore::SVGPathBlender::blendArcToSegment):
* Source/WebCore/svg/SVGPathByteStreamSource.cpp:
(WebCore::SVGPathByteStreamSource::parseMoveToSegment):
(WebCore::SVGPathByteStreamSource::parseLineToSegment):
(WebCore::SVGPathByteStreamSource::parseLineToHorizontalSegment):
(WebCore::SVGPathByteStreamSource::parseLineToVerticalSegment):
(WebCore::SVGPathByteStreamSource::parseCurveToCubicSegment):
(WebCore::SVGPathByteStreamSource::parseCurveToCubicSmoothSegment):
(WebCore::SVGPathByteStreamSource::parseCurveToQuadraticSegment):
(WebCore::SVGPathByteStreamSource::parseCurveToQuadraticSmoothSegment):
(WebCore::SVGPathByteStreamSource::parseArcToSegment):
* Source/WebCore/svg/SVGPathByteStreamSource.h:
* Source/WebCore/svg/SVGPathParser.cpp:
(WebCore::SVGPathParser::parseMoveToSegment):
(WebCore::SVGPathParser::parseLineToSegment):
(WebCore::SVGPathParser::parseLineToHorizontalSegment):
(WebCore::SVGPathParser::parseLineToVerticalSegment):
(WebCore::SVGPathParser::parseCurveToCubicSegment):
(WebCore::SVGPathParser::parseCurveToCubicSmoothSegment):
(WebCore::SVGPathParser::parseCurveToQuadraticSegment):
(WebCore::SVGPathParser::parseCurveToQuadraticSmoothSegment):
(WebCore::SVGPathParser::parseArcToSegment):
* Source/WebCore/svg/SVGPathSegListSource.cpp:
(WebCore::SVGPathSegListSource::parseMoveToSegment):
(WebCore::SVGPathSegListSource::parseLineToSegment):
(WebCore::SVGPathSegListSource::parseLineToHorizontalSegment):
(WebCore::SVGPathSegListSource::parseLineToVerticalSegment):
(WebCore::SVGPathSegListSource::parseCurveToCubicSegment):
(WebCore::SVGPathSegListSource::parseCurveToCubicSmoothSegment):
(WebCore::SVGPathSegListSource::parseCurveToQuadraticSegment):
(WebCore::SVGPathSegListSource::parseCurveToQuadraticSmoothSegment):
(WebCore::SVGPathSegListSource::parseArcToSegment):
* Source/WebCore/svg/SVGPathSegListSource.h:
* Source/WebCore/svg/SVGPathSource.h:
* Source/WebCore/svg/SVGPathStringViewSource.cpp:
(WebCore::SVGPathStringViewSource::parseMoveToSegment):
(WebCore::SVGPathStringViewSource::parseLineToSegment):
(WebCore::SVGPathStringViewSource::parseLineToHorizontalSegment):
(WebCore::SVGPathStringViewSource::parseLineToVerticalSegment):
(WebCore::SVGPathStringViewSource::parseCurveToCubicSegment):
(WebCore::SVGPathStringViewSource::parseCurveToCubicSmoothSegment):
(WebCore::SVGPathStringViewSource::parseCurveToQuadraticSegment):
(WebCore::SVGPathStringViewSource::parseCurveToQuadraticSmoothSegment):
(WebCore::SVGPathStringViewSource::parseArcToSegment):
* Source/WebCore/svg/SVGPathStringViewSource.h:

Canonical link: <a href="https://commits.webkit.org/285347@main">https://commits.webkit.org/285347@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/35d73f5211040b160814056cfde5c343d78dbd3f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/72253 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/51674 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/25041 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/76422 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/23462 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/74368 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/59478 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/23284 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/56959 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15455 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/75320 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/46812 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/62223 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/37379 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/43469 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/19698 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/21812 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/65366 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/20058 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/78098 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/16494 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/19205 "Passed tests") | [⏳ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/GTK-WK2-Tests-EWS "Waiting to run tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/16541 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/62247 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15980 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/12904 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/6544 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/47472 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/2256 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/48541 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/49829 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/48284 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Source/WebCore/svg/SVGPathStringViewSource.h:
```diff
@@ -36,15 +36,15 @@ class SVGPathStringViewSource final : public SVGPathSource {
-    std::optional<CurveToQuadraticSegment> parseCurveToQuadraticSegment() final;
-    std::optional<CurveToQuadraticSmoothSegment> parseCurveToQuadraticSmoothSegment() final;
-    std::optional<ArcToSegment> parseArcToSegment() final;
+    std::optional<MoveToSegment> parseMoveToSegment(FloatPoint) final;
```
Should we add `currentPoint` as the name of the FloatPoint?

smfr replied:
It's named in pure virtual base class interface, which I think is the right place to name it.

smfr committed:
Pass the current point to SVGPathSource segment parsing functions
https://bugs.webkit.org/show_bug.cgi?id=281645
rdar://138088365

Reviewed by Alan Baradlay.

The CSS `shape()` function is getting start/end-relative control points[1] and so its SVGPathSource
implementation will need to know the current path point. So pass the current point to all SVGPathSource
functions that create non-trivial segments.

[1] https://github.com/w3c/csswg-drafts/issues/10649#issuecomment-2412816773

* Source/WebCore/rendering/style/BasicShapesShape.cpp:
* Source/WebCore/svg/SVGPathBlender.cpp:
(WebCore::pullFromSources):
(WebCore::SVGPathBlender::blendMoveToSegment):
(WebCore::SVGPathBlender::blendLineToSegment):
(WebCore::SVGPathBlender::blendLineToHorizontalSegment):
(WebCore::SVGPathBlender::blendLineToVerticalSegment):
(WebCore::SVGPathBlender::blendCurveToCubicSegment):
(WebCore::SVGPathBlender::blendCurveToCubicSmoothSegment):
(WebCore::SVGPathBlender::blendCurveToQuadraticSegment):
(WebCore::SVGPathBlender::blendCurveToQuadraticSmoothSegment):
(WebCore::SVGPathBlender::blendArcToSegment):
* Source/WebCore/svg/SVGPathByteStreamSource.cpp:
(WebCore::SVGPathByteStreamSource::parseMoveToSegment):
(WebCore::SVGPathByteStreamSource::parseLineToSegment):
(WebCore::SVGPathByteStreamSource::parseLineToHorizontalSegment):
(WebCore::SVGPathByteStreamSource::parseLineToVerticalSegment):
(WebCore::SVGPathByteStreamSource::parseCurveToCubicSegment):
(WebCore::SVGPathByteStreamSource::parseCurveToCubicSmoothSegment):
(WebCore::SVGPathByteStreamSource::parseCurveToQuadraticSegment):
(WebCore::SVGPathByteStreamSource::parseCurveToQuadraticSmoothSegment):
(WebCore::SVGPathByteStreamSource::parseArcToSegment):
* Source/WebCore/svg/SVGPathByteStreamSource.h:
* Source/WebCore/svg/SVGPathParser.cpp:
(WebCore::SVGPathParser::parseMoveToSegment):
(WebCore::SVGPathParser::parseLineToSegment):
(WebCore::SVGPathParser::parseLineToHorizontalSegment):
(WebCore::SVGPathParser::parseLineToVerticalSegment):
(WebCore::SVGPathParser::parseCurveToCubicSegment):
(WebCore::SVGPathParser::parseCurveToCubicSmoothSegment):
(WebCore::SVGPathParser::parseCurveToQuadraticSegment):
(WebCore::SVGPathParser::parseCurveToQuadraticSmoothSegment):
(WebCore::SVGPathParser::parseArcToSegment):
* Source/WebCore/svg/SVGPathSegListSource.cpp:
(WebCore::SVGPathSegListSource::parseMoveToSegment):
(WebCore::SVGPathSegListSource::parseLineToSegment):
(WebCore::SVGPathSegListSource::parseLineToHorizontalSegment):
(WebCore::SVGPathSegListSource::parseLineToVerticalSegment):
(WebCore::SVGPathSegListSource::parseCurveToCubicSegment):
(WebCore::SVGPathSegListSource::parseCurveToCubicSmoothSegment):
(WebCore::SVGPathSegListSource::parseCurveToQuadraticSegment):
(WebCore::SVGPathSegListSource::parseCurveToQuadraticSmoothSegment):
(WebCore::SVGPathSegListSource::parseArcToSegment):
* Source/WebCore/svg/SVGPathSegListSource.h:
* Source/WebCore/svg/SVGPathSource.h:
* Source/WebCore/svg/SVGPathStringViewSource.cpp:
(WebCore::SVGPathStringViewSource::parseMoveToSegment):
(WebCore::SVGPathStringViewSource::parseLineToSegment):
(WebCore::SVGPathStringViewSource::parseLineToHorizontalSegment):
(WebCore::SVGPathStringViewSource::parseLineToVerticalSegment):
(WebCore::SVGPathStringViewSource::parseCurveToCubicSegment):
(WebCore::SVGPathStringViewSource::parseCurveToCubicSmoothSegment):
(WebCore::SVGPathStringViewSource::parseCurveToQuadraticSegment):
(WebCore::SVGPathStringViewSource::parseCurveToQuadraticSmoothSegment):
(WebCore::SVGPathStringViewSource::parseArcToSegment):
* Source/WebCore/svg/SVGPathStringViewSource.h:

Canonical link: https://commits.webkit.org/285347@main

shallawa commented on Source/WebCore/svg/SVGPathParser.cpp:
```diff
@@ -94,7 +94,7 @@ bool SVGPathParser::parseMoveToSegment()
 bool SVGPathParser::parseLineToSegment()
 {
-    auto result = m_source->parseLineToSegment();
+    auto result = m_source->parseLineToSegment(m_currentPoint);
```
I do not understand fully the purpose of this change but I expect it is work towards other clean-up and more implementation. With this change this function becomes a little bit confusing. We are passing `m_currentPoint` to `m_source-> parseLineToSegment()`. But if `(m_pathParsingMode == NormalizedParsing` is true we are going to add `m_currentPoint` to the `result->targetPoint`. So I would expect the next step will be to make the `SVGPathSource` handle the addition of `m_currentPoint` to the result in the case of `(m_pathParsingMode == NormalizedParsing`
```
bool SVGPathParser::parseLineToSegment()
{
    auto result = m_source->parseLineToSegment(m_currentPoint);
    if (!result)
        return false;

    if (m_pathParsingMode == NormalizedParsing)
        m_consumer->lineTo(m_currentPoint, AbsoluteCoordinates);
    else
        m_consumer->lineTo(result->targetPoint, m_mode);

     m_currentPoint = result->targetPoint;
    return true;
}
```
And of course `SVGPathSource` has to be aware of `m_pathParsingMode` and `m_mode`.
